Step4 계산기 기능 UI테스트

oyj7677 opened the pull request:
step3와 관련된 개선 사항과 함께 step4 구현 PR합니다. 감사합니다.

Commit:
feat: 사용자가 피연산자 0 ~ 9 버튼을 누르면 화면에 해당 숫자를 보여준다

Commit:
Merge branch 'main' of https://github.com/oyj7677/android-calculator-mvp

Commit:
Merge branch 'main' of https://github.com/oyj7677/android-calculator-mvp

Commit:
step3 개선 사항
- ParameterizedTest Class 분리
- 코드 가독성 개선
	- kotlin.runCatching{ } → runCatching{ }
	- String.require() 사용 및 isEmpty확장함수 사용
- enum 활용 개선
	- Operator객체 반환하여 사용

Commit:
step3 개선 사항
- Operator 계산 전략 enum에 귀속
- when절 삭제

Commit:
step 4 계산기 기능 UI 테스트

oyj7677 commented:
step 4 관련 PR입니다. 감사합니다.

oyj7677 commented:
수식에 대해 지우기 버튼을 다수 누르는 테스트를 할 때, 지우기 버튼 클릭마다 해당 값을 검증하는 코드를 하나의 테스트 클래스로 작성할 수 있을까요? 
ex) 32 + 1 -> 지우기 버튼 -> 32 + 검증 -> 지우기 버튼 -> 32 검증 -> 지우기 버튼 -> 3 검증 -> 지우기 버튼 -> 공백 검증
총 4번의 검증을 하나의 테스트 함수에서 처리하고 싶습니다.

malibinYun commented on app/src/androidTest/java/camp/nextstep/edu/calculator/MainActivityTest.kt:
```diff
@@ -13,6 +13,121 @@ class MainActivityTest {
+        // then : 연결된 피연산자의 문자가 보여야 한다.
+        onView(withId(R.id.textView)).check(matches(withText("17")))
+    }
+
```
UI 테스트를 잘 작성해주셨네요! 👍
버튼을 누른 것만 보면 어떤 수식인지 한 눈에 보기 어려우니, 주석으로 어떤 값이 입력되었는지 적어주시면 더 가독성을 챙길 수 있게 됩니다 :)

```
// given : 5 +
...

malibinYun commented on app/src/androidTest/java/camp/nextstep/edu/calculator/MainActivityTest.kt:
```diff
@@ -13,6 +13,121 @@ class MainActivityTest {
+        onView(withId(R.id.buttonDelete)).perform(click()) //
+
+        // than : 마지막 입력된 항목이 삭제 되고 화면에 보여야 한다.
+        onView(withId(R.id.textView)).check(matches(withText("")))
```
> 수식에 대해 지우기 버튼을 다수 누르는 테스트를 할 때, 지우기 버튼 클릭마다 해당 값을 검증하는 코드를 하나의 테스트 클래스로 작성할 수 있을까요?
ex) 32 + 1 -> 지우기 버튼 -> 32 + 검증 -> 지우기 버튼 -> 32 검증 -> 지우기 버튼 -> 3 검증 -> 지우기 버튼 -> 공백 검증
총 4번의 검증을 하나의 테스트 함수에서 처리하고 싶습니다.

말씀하신대로 구현하셔도 좋습니다!
이런 경우에는 when / then 이 여러 번 나타나는 구조가 되겠지요 :)

// when
onView(withId(R.id.buttonDelete)).perform(click())
// then
onView(withId(R.id.textView)).check(matches(withText("32 +")))

//when
...

malibinYun commented on app/src/main/java/camp/nextstep/edu/calculator/MainActivity.kt:
```diff
@@ -1,27 +1,97 @@
+        binding.buttonMultiply.setOnClickListener{addOperator('*')}
+        binding.buttonDivide.setOnClickListener{addOperator('/')}
+        binding.buttonDelete.setOnClickListener{deleteLastStr()}
+        binding.buttonEquals.setOnClickListener{calculate()}
```
코드 자동 정렬 단축키를 활용해서 정렬해보는 건 어떨까요?
윈도우 단축키는 `ctrl + alt + L` 이고,
맥은 `command + option + L` 입니다 :)

malibinYun commented on app/src/main/java/camp/nextstep/edu/calculator/MainActivity.kt:
```diff
@@ -1,27 +1,97 @@
 
+        numAndOperator = numAndOperator.dropLast(1)
+        numAndOperator = numAndOperator.trim()
+        showText(numAndOperator)
```
"수식" 이란 도메인 객체를 만들어보는 건 어떨까요?
calculator 처럼 순수한 코틀린 로직으로도 충분히 만들어볼 수 있어 보입니다 :)
순수 코틀린으로 만들어보시고, domain 모듈에 위치 시켜 보세요 🙂

malibinYun commented on app/src/test/java/camp/nextstep/edu/calculator/CalculatorTest.kt:
```diff
@@ -64,31 +74,48 @@ class CalculatorTest(val input: String, val expected: Int) {
     }
 
 
+    @Test
+    fun 여러_에러_케이스를_테스트한다() {
+        // when : 에러가 발생하는 값들을 넣는다.
+        val exception = runCatching { calculator.evaluate(input) }.exceptionOrNull()
+
+        // then : IllegalArgumentException이 발생한다.
+        assertThat(exception).isInstanceOf(expected::class.java)
+    }
```
여러 케이스의 테스트가 필요한 것만 class 분리를 해보시라는 의미었습니다.
해당 클래스에서는 멤버변수로 제공되는 testData()를 이 테스트 메서드를 제외하고는 사용하지 않기 때문에 적절히 묶여있다고 보기 어려워 보여요 :)

위의 CalculatorTest 도 같은 맥락입니다.

지금의 경우에는, 테스트 클래스는 3개로 분리될 수 있겠지요!
1개 클래스에는 파라미터 관련한 멤버 변수가 존재하지 않게 되고,
2개의 클래스는 각각 한 개의 테스트 메서드를 갖게 되겠네요 :)

malibinYun commented on calculatorLib/src/main/java/com/example/calculatorlib/Calculator.kt:
```diff
@@ -3,54 +3,31 @@ package com.example.calculatorlib
+
+        require(!str.isEmpty()) {
+            "문자열 수식(str)이 공백일 경우 IllegalArgumentException 발행"
+        }
```
공백인 경우는 이미 isNullOrEmpty로부터 검증이 되어서 불필요해 보이네요 🙂

malibinYun commented on calculatorLib/src/main/java/com/example/calculatorlib/Operator.kt:
```diff
@@ -4,5 +4,29 @@ enum class Operator(val operator: String) {
-    DIVIDED("/")
+    DIVIDED("/");
+
+    fun calculator(result: Int, num: Int) = when(Operator.find(operator)) {
```
```kotlin
enum class Operator(val symbol: String, calculate: (Int, Int) -> Int) {
    PLUS("+", { left, right -> left + right }),
}
```
위 처럼 enum 항목이 직접 계산 전략을 담은 람다를 갖게 만들 수 있어요!
그러면 https://github.com/next-step/android-calculator-mvp/pull/196#discussion_r1278705057 이 피드백에서 드렸던 마지막 내용을 분기문 없이 구현해볼 수 있습니다 :)

malibinYun commented on calculatorLib/src/main/java/com/example/calculatorlib/Operator.kt:
```diff
@@ -4,5 +4,29 @@ enum class Operator(val operator: String) {
+        DIVIDED -> division(result, num)
+    }
+    companion object {
+        fun find(operator: String): Operator {
+            return values().find { it.operator == operator } ?: throw IllegalArgumentException("사칙연산 기호가 아닌 경우 IllegalArgumentException 발행")
+        }
+
```
어떤 값이 들어와서 Exception이 발생했는지 메세지에 넣어 보여주는 건 어떨까요?
적용했을 때 실제 Exception은 로그 창에서 어떻게 나타나는 지 확인해보시면 좋겠네요! :)

malibinYun reviewed (changes requested):
4단계 미션 고생 많으셨습니다!
생각해보면 좋을 점들에 코멘트 남겨두었습니다.
피드백 반영 후 다시 리뷰 요청 부탁드려요!

Commit:
step3, step4 코드 개선
 - UI테스트 입력값 주석으로 표시
 - 하나의 클래스에 when - then 반복 구성
 - Parameterized 테스트 클래스 분리
 - 코드 정리
 - 수식(formula)모듈 생성
 - enum에 람다식 사용
 - 에러표시 개선

oyj7677 commented:
step3, step4 코드 개선
- UI테스트 입력값 주석으로 표시
- 하나의 클래스에 when - then 반복 구성
- Parameterized 테스트 클래스 분리
- 코드 정리
- 수식(formula)모듈 생성
- enum에 람다식 사용
- 에러표시 개선

malibinYun commented on FormulaLib/src/main/java/com/example/formulalib/Formula.kt:
```diff
@@ -0,0 +1,58 @@
+package com.example.formulalib
+
+class Formula {
```
문자열 계산기라는 도메인으로 묶을 수 있어 보여요 :)
기존에 존재하던 calculatorLib은 domain으로 이름을 변경하고,
수식은 기존에 존재하던 domain 모듈에 위치시키는 건 어떨까요?

malibinYun replied:
또, 수식에 대한 유닛 테스트를 작성해보는 건 어떨까요?

malibinYun commented on app/src/test/java/camp/nextstep/edu/calculator/MainActivityLogicTest.kt:
```diff
@@ -0,0 +1,35 @@
+import com.google.common.truth.Truth.assertThat
+import org.junit.Test
+
+class MainActivityLogicTest {
+
+    private var numAndOperator = "12 +"
+
+    @Test
+    fun 마지막_값을_검증한다() {
+        // given: 문자열의 마지막 인덱스를 구한다.
+        val lastIndex = numAndOperator.length - 1
+
+        // when: 문자열의 마지막 값이 숫자인지 확인합니다.
+        val actual = numAndOperator[lastIndex].isDigit()
+
+        // then: 숫자일 경우 true, 숫자 이외의 것 이라면 false
+        assertThat(actual).isEqualTo(false)
+    }
+
+    @Test
+    fun 마지막으로_입력된_연산자를_교체한다() {
+        // given: 문자열의 마지막 인덱스를 구한다.
+        val lastIndex = numAndOperator.length - 1
+
+        // when: 마지막 문자열의 값을 삭제 후 추가한다.
+        numAndOperator = numAndOperator.dropLast(1)
+        numAndOperator += "*"
+        val actual = numAndOperator[lastIndex]
+
+        // then: 숫자일 경우 true, 숫자 이외의 것 이라면 false
+        assertThat(actual).isEqualTo('*')
+    }
+}
```
이 테스트는 정말 MainActivity를 테스트하는 로직이 맞을까요?
mainActivity의 로직이 변경되어서 에러가 생기더라도, 이 테스트는 여전히 통과하는 테스트입니다.
그렇다면 이 테스트는 거짓말을 하게 되는 테스트가 되는데요, 이런 경우 이 테스트의 유지보수가 어려워집니다.

현재는 수식이라는 도메인 객체가 생겨났으니, 해당 객체를 테스트해봄으로써 원하던 테스트 해볼 수 있겠네요 :)

malibinYun reviewed (approved):
4단계 미션 고생 많으셨습니다 :)
마지막 미션을 앞두고 계시네요! 마지막까지 응원하겠습니다 :)
다음 미션을 위해 머지할게요!
피드백은 다음 미션에 반영해주세요 🙂

malibinYun commented:
기능 단위로 커밋을 쪼개보는 건 어떨까요?
1개의 커밋의 부피가 커지면, 변경점을 알아보기 힘들어집니다.

기능 단위로 쪼개 커밋하는 것을 연습해보는 좋은 기회가 될 것이라 생각해요!
